### ISSUE TITLE
Missing wait-states before DPP instructions (v_mov_dpp)

### ISSUE THREAD

**Issue by ex-rzr** [RadeonOpenCompute/hcc#715]
We are developing dpp-based reduction and scan for https://github.com/ROCmSoftwarePlatform/rocPRIM and some test cases are failing (all of them work with permute-based reduction and scan).

DPP instructions require 2 wait states after instructions that write into VGPRs read by the DPP:
http://developer.amd.com/wordpress/media/2017/08/Vega_Shader_ISA_28July2017.pdf ("4.5. Manually Inserted Wait States (NOPs)").

https://gpuopen.com/amd-gcn-assembly-cross-lane-operations/ says: 
> When using these instrinsics, HCC will automatically ensure that the proper .waitcnt and data dependencies are honored.

Currently we use `llvm.amdgcn.update.dpp.i32` instead of `__amdgcn_move_dpp`
(**no inline asm**) and the compiler indeed ensures that, but sometimes there is only one wait state (I didn't see zero wait states but I'm not sure this never happens).

With `llvm.amdgcn.move.dpp` more tests fail (IIRC this intrisic is obsolete, and `llvm.amdgcn.update.dpp` should be used).

Perhaps the compiler inserts all required wait states correctly but its next stages remove them.

Here are ISAs for some of failing test cases:
```
test_hip_device_segmented_scan-gfx900.isa-  v_add_u32_e32 v1, v1, v11                                  // 00000004CF80: 68021701 
test_hip_device_segmented_scan-gfx900.isa-  v_cndmask_b32_e32 v2, v2, v7, vcc                          // 00000004CF84: 00040F02 
test_hip_device_segmented_scan-gfx900.isa:  v_mov_b32_dpp v13, v1  quad_perm:[1,0,3,2] row_mask:0xf bank_mask:0xf// 00000004CF88: 7E1A02FA FF00B101 

test_hip_device_segmented_scan-gfx900.isa-  v_add_u32_e32 v5, v5, v11                                  // 000000050A9C: 680A1705 
test_hip_device_segmented_scan-gfx900.isa-  v_cndmask_b32_e32 v1, v1, v6, vcc                          // 000000050AA0: 00020D01 
test_hip_device_segmented_scan-gfx900.isa:  v_mov_b32_dpp v16, v5  quad_perm:[1,0,3,2] row_mask:0xf bank_mask:0xf// 000000050AA4: 7E2002FA FF00B105

test_hip_device_segmented_scan-gfx900.isa-  v_add_u32_e32 v5, v5, v11                                  // 000000052584: 680A1705 
test_hip_device_segmented_scan-gfx900.isa-  v_cndmask_b32_e32 v1, v1, v6, vcc                          // 000000052588: 00020D01 
test_hip_device_segmented_scan-gfx900.isa:  v_mov_b32_dpp v15, v5  quad_perm:[1,0,3,2] row_mask:0xf bank_mask:0xf// 00000005258C: 7E1E02FA FF00B105 

test_hip_device_segmented_scan-gfx900.isa-  v_add_u32_e32 v5, v5, v11                                  // 000000053184: 680A1705 
test_hip_device_segmented_scan-gfx900.isa-  v_cndmask_b32_e32 v1, v1, v6, vcc                          // 000000053188: 00020D01 
test_hip_device_segmented_scan-gfx900.isa:  v_mov_b32_dpp v15, v5  quad_perm:[1,0,3,2] row_mask:0xf bank_mask:0xf// 00000005318C: 7E1E02FA FF00B105 

test_hip_device_segmented_scan-gfx803.isa-  v_add_u32_e32 v8, vcc, v39, v8                             // 000000067668: 32101127 
test_hip_device_segmented_scan-gfx803.isa-  v_bfe_i32 v27, v18, 4, 1                                   // 00000006766C: D1C9001B 02050912 
test_hip_device_segmented_scan-gfx803.isa:  v_mov_b32_dpp v24, v8  row_shr:1 row_mask:0xf bank_mask:0xf// 000000067674: 7E3002FA FF011108

test_hip_device_segmented_scan-gfx900.isa-  v_add_u32_e32 v7, v7, v38                                  // 00000005C220: 680E4D07 
test_hip_device_segmented_scan-gfx900.isa-  s_nop 0                                                    // 00000005C224: BF800000 
test_hip_device_segmented_scan-gfx900.isa:  v_mov_b32_dpp v22, v7  row_shr:1 row_mask:0xf bank_mask:0xf// 00000005C228: 7E2C02FA FF011107 

test_hip_device_segmented_scan-gfx900.isa-  v_add_f32_e32 v19, v19, v40                                // 000000061B68: 02265113 
test_hip_device_segmented_scan-gfx900.isa-  s_nop 0                                                    // 000000061B6C: BF800000 
test_hip_device_segmented_scan-gfx900.isa:  v_mov_b32_dpp v20, v19  row_shr:1 row_mask:0xf bank_mask:0xf// 000000061B70: 7E2802FA FF011113 

test_hip_device_segmented_scan-gfx900.isa-  v_max3_i32 v18, v18, v38, v39                              // 00000006335C: D1D40012 049E4D12 
test_hip_device_segmented_scan-gfx900.isa-  s_nop 0                                                    // 000000063364: BF800000 
test_hip_device_segmented_scan-gfx900.isa:  v_mov_b32_dpp v19, v18  row_shr:1 row_mask:0xf bank_mask:0xf// 000000063368: 7E2602FA FF011112

test_hip_device_segmented_scan-gfx900.isa-  v_add_u32_e32 v18, v18, v39                                // 000000064B68: 68244F12 
test_hip_device_segmented_scan-gfx900.isa-  s_nop 0                                                    // 000000064B6C: BF800000 
test_hip_device_segmented_scan-gfx900.isa:  v_mov_b32_dpp v19, v18  row_shr:1 row_mask:0xf bank_mask:0xf// 000000064B70: 7E2602FA FF011112 

test_hip_device_segmented_scan-gfx900.isa-  v_add_u32_e32 v8, v8, v39                                  // 000000066438: 68104F08 
test_hip_device_segmented_scan-gfx900.isa-  s_nop 0                                                    // 00000006643C: BF800000 
test_hip_device_segmented_scan-gfx900.isa:  v_mov_b32_dpp v23, v8  row_shr:1 row_mask:0xf bank_mask:0xf// 000000066440: 7E2E02FA FF011108
```

**There is a chance that the compiler fails to add correct number of wait states before other instructions listed in "4.5. Manually Inserted Wait States (NOPs)"**.

I think this is very dangerous because such situations (at least in our case) are very rare and extremely hard to localize even with tests.

**Comment by b-sumner** [comment]
Would you be able to provide (or point to) some minimal examples that we can feed through the compiler to reproduce the problem?

**Comment by ex-rzr** [comment]
@b-sumner 

Here is a minimized test based on our real code: https://gist.github.com/ex-rzr/fbd364d79b6a84a7f1abc624f12343c8

```sh
hipcc -o dpp_bug dpp_bug.cpp && ./dpp_bug
```

It seems that the bug appears only in certain conditions, that's why there are all these weird shared memory and sync operations. Sometimes the ISA is correct when I change the code even slightly. For example, it works with `ItemsPerThread = 4`.

It may be a coincidence but I've never seen this issue for other dpp, only for the first one.

How it looks like:
gfx900 (MI25)
```
        v_add_u32_e32 v2, v2, v8                                   // 0000000011D0: 68041102 
        s_nop 0                                                    // 0000000011D4: BF800000 
        v_mov_b32_dpp v9, v2  quad_perm:[1,0,3,2] row_mask:0xf bank_mask:0xf// 0000000011D8: 7E1202FA FF00B102  
```

gfx803 (9300)
```
        v_add_u32_e32 v2, vcc, v8, v2                              // 0000000011D4: 32040508 
        s_nop 0                                                    // 0000000011D8: BF800000 
        v_mov_b32_dpp v9, v2  quad_perm:[1,0,3,2] row_mask:0xf bank_mask:0xf// 0000000011DC: 7E1202FA FF00B102 
```
(By the way, 9300 produces correct reduction results when the ISA is incorrect for some versions of the code).

Tested on
```
HIP version: 1.5.18081
HCC clang version 6.0.0 (ssh://gerritgit/compute/ec/hcc-tot/clang d44cba95b872ae463c2dd48131543bdeae81be5c) (ssh://gerritgit/compute/ec/hcc-tot/llvm c3a2467ca4a3a8498e89ebf787e9a947d53a15cb) (based on HCC 1.2.18142-588ad64-d44cba9-c3a2467 )
```


`llvm.amdgcn.update.dpp.i32` is used in the test, but there is a commented line with `hc::__amdgcn_move_dpp` call, you may check it too. Both instructions are affected by this issue. However in our case `hc::__amdgcn_move_dpp` fails in much more rocPRIM tests than `llvm.amdgcn.update.dpp.i32` (and I suppose this is another bug).
The commit message here https://github.com/llvm-mirror/llvm/commit/8e92b2817eaf9d068f5ef86584b6043f6a1ea4a3 mentions that `llvm.amdgcn.mov.dpp.i32` should be deprecated.

**Comment by b-sumner** [comment]
@ex-rzr thanks for the example and the discussion about it.  We'll take a look.

BTW,  I see the code is setting old=0 and bount_ctrl = false.  You can get the same effect and potentially better code by setting bound_ctrl = true (and old to anything including input).

**Comment by ex-rzr** [comment]
> BTW, I see the code is setting old=0 and bount_ctrl = false. You can get the same effect and potentially better code by setting bound_ctrl = true (and old to anything including input).

True, this feature can be useful for scan on primitive types with functions like sum, min, max, and, or. And here (reduction) using input as `old` may potentially save one `v_mov` (this needs checking), because we don't care about values except the last of the "warp", i.e. `WarpSize - 1`.

However we have only generic implementations of scan and reduce for any arbitrary scan/reduction operations for now (because correctness is more important).

**Comment by searlmc1** [comment]
From an earlier comment: "It may be a coincidence but I've never seen this issue for other dpp, only for the first one."

It does not look like coincidence per se. I've included some IR snippets, but what I see when looking at the IR for dpp_bug.cpp is that:
- for all but the first dpp, post-ra scheduler generates 2 S_NOPs to satisfy the 2 wait state requirement
- for the first dpp, post-ra scheduler schedules an ATOMIC_FENCE and generates a S_NOP to satisfy the 2 wait state requirement 
- per the memory legalizer, no instrs are needed when expanding the ATOMIC_FENCE, which means there is now but a single S_NOP / single wait state, which is wrong

IR Dump Before Post RA top-down list latency scheduler
…
  renamable $vgpr2 = V_ADD_U32_e32 killed $vgpr2, killed $vgpr8, implicit $exec
  renamable $vgpr9 = V_MOV_B32_dpp killed $vgpr9, $vgpr2, 177, 15, 15, 0, implicit $exec
…

IR Dump After Post RA top-down list latency scheduler
…
  renamable $vgpr2 = V_ADD_U32_e32 killed $vgpr2, killed $vgpr8, implicit $exec
  ATOMIC_FENCE 4, 3
  S_NOP 0
  renamable $vgpr9 = V_MOV_B32_dpp killed $vgpr9, $vgpr2, 177, 15, 15, 0, implicit $exec
…

IR Dump Before SI Memory Legalizer
…
  renamable $vgpr2 = V_ADD_U32_e32 killed $vgpr2, killed $vgpr8, implicit $exec
  ATOMIC_FENCE 4, 3
  S_NOP 0
  renamable $vgpr9 = V_MOV_B32_dpp killed $vgpr9, $vgpr2, 177, 15, 15, 0, implicit $exec
…

IR Dump After SI Memory Legalizer
…
  renamable $vgpr2 = V_ADD_U32_e32 killed $vgpr2, killed $vgpr8, implicit $exec
  S_NOP 0
  renamable $vgpr9 = V_MOV_B32_dpp killed $vgpr9, $vgpr2, 177, 15, 15, 0, implicit $exec
…

**Comment by searlmc1** [comment]
Fixed in the llvm-trunk via https://reviews.llvm.org/rL337154

**Comment by jszuppe** [comment]
Is it fixed in recent ROCm release?

**Comment by searlmc1** [comment]
Hi,

This was fixed in ROCm 1.9.x, though I would suggest using one of the more recent ROCm 2.x releases.

Let us know if you run into any issues.
